cpp/arcticdb/async/tasks.hpp

```diff
@@ -492,24 +492,54 @@ struct DecodeMetadataAndDescriptorTask : BaseTask {
             );
     }
 };
+template<class...>
+using voider = void;
+template <typename Iter, typename U, typename = void>
+struct is_iter_of : std::false_type{};
 
-template<typename ConstVarKeyGetter,
-        typename=std::enable_if_t<!std::is_reference_v<ConstVarKeyGetter> // Make obvious if iterator& is captured
-                        && std::is_same_v<decltype(*std::declval<ConstVarKeyGetter>()), const entity::VariantKey&>>>
+template <typename Iter, typename U>
+struct is_iter_of<Iter, U, voider<decltype(*(std::declval<Iter>()))>> : std::is_constructible<U, decltype(*(std::declval<Iter>()))> {};
+
+template<typename T, typename U>
+struct is_pointer_of : std::false_type {};
+
+template<typename T, typename U>
+struct is_pointer_of<T*, U> : std::is_same<std::remove_pointer_t<T*>, U> {};
+
+template<typename T, typename U>
+using is_same_cvref_removed = std::is_same<std::remove_cvref_t<T>, U>;
+
+template<typename T,
+        typename=std::enable_if_t< // Make obvious if iterator& is captured
+                        is_same_cvref_removed<T, entity::VariantKey>::value
+                        || is_iter_of<T, entity::VariantKey>::value
+                        || is_pointer_of<T, entity::VariantKey>::value>>
 struct KeyExistsTask : BaseTask {
```

Comment:
The old implementation was going to great lengths to ensure that we are never copying the key in to the `KeyExistsTask`, why do we need to relax that constraint now??

Reply:
In the old implementation, the task itself doesn't hold the memory during construction. This could be a ticking timebomb (at least I hit it) as the memory could no longer be valid by the time the task is run.
IMO in the worker-pool model, every task should be self-sufficient - not depending on memory resides outside the class

Reply:
This is still dangerous, it is allowing iterators into vectors that can go out of scope and raw pointers. To ensure liveness of the key memory, only accept rvalue references and shared pointers

Reply:
I am happy to do so if there is no concern for performance. It can simplify the mad template parameter checking in this class.
I thought it is designed to accept iterator and shared pointer for a reason.
My changes is only for making the class store the key locally if a lvalue key is passed

Reply:
Sorry, yes, copying lvalues is also acceptable from a correctness perspective. I haven't seen this level of madness in any similar methods though, how do they solve this problem? I've definitely seen overloaded constructors that copy lvalues or move rvalues, are there other approaches?

Reply:
The template has been removed as `KeyExistsTask` now store copy of key, per discussed